pkg/image/resolve.go

```diff
@@ -61,20 +59,72 @@ func getChartAndVersion(path string) (map[string]chart, error) {
 	if err != nil {
 		return nil, err
 	}
-	for k, versions := range index.IndexFile.Entries {
-		// because versions is sorted in reverse order, the first one will be the latest version
+	for _, versions := range index.IndexFile.Entries {
 		if len(versions) > 0 {
-			newestVersionedChart := versions[0]
-			rtn[k] = chart{
-				dir:     newestVersionedChart.Dir,
-				version: newestVersionedChart.Version}
+			// Add latest version (First element is latest, list sorted in descending order)
+			chartVersions = append(chartVersions, versions[0])
+			if strings.Contains(path, SystemChartsRepoDir) {
+				// Skip if rancherVersion is not a valid semver (E.g is a commit hash)
+				if rancherSemVer, err := semver.NewVersion(rancherVersion); err == nil {
+					// Skip first element, the latest version has already been added
+					for _, v := range versions[1:] {
+						questions, err := getQuestions(filepath.Join(path, v.Dir))
+						if err != nil {
+							return nil, fmt.Errorf("no questions file in chart %s:%s", v.Name, v.Version)
+						}
+						min, ok := questions["rancher_min_version"].(string)
+						if !ok {
+							return nil, fmt.Errorf("no rancher_min_version set in chart %s:%s", v.Name, v.Version)
+						}
+						// No ok check, max is optional
+						max, _ := questions["rancher_max_version"].(string)
+						isInRange, err := isInMinMaxRange(rancherSemVer, min, max)
+						if err != nil {
+							return nil, err
+						}
+						if isInRange {
+							chartVersions = append(chartVersions, v)
+						}
+					}
+				}
+			}
 		}
 	}
+	return chartVersions, nil
+}
 
-	return rtn, nil
+func getQuestions(versionPath string) (map[interface{}]interface{}, error) {
+	content, err := ioutil.ReadFile(filepath.Join(versionPath, "questions.yaml"))
+	if err != nil {
+		content, err = ioutil.ReadFile(filepath.Join(versionPath, "questions.yml"))
+		if err != nil {
+			return nil, err
+		}
+	}
+	questions := make(map[interface{}]interface{})
+	if err = yaml.Unmarshal(content, &questions); err != nil {
+		return nil, err
+	}
+	return questions, nil
+}
+
+func isInMinMaxRange(rancherSemVer *semver.Version, min, max string) (bool, error) {
+	var minSemVer, maxSemVer *semver.Version
+	var isInRange bool
+	var err error
+	if minSemVer, err = semver.NewVersion(strings.TrimSpace(min)); err == nil {
+		if maxSemVer, err = semver.NewVersion(strings.TrimSpace(max)); err == nil {
+			isInRange = (rancherSemVer.GreaterThan(minSemVer) || rancherSemVer.Equal(minSemVer)) &&
```

Comment:
Not less than is equivalent to `greatherThanOrEqual`:
```suggestion
			isInRange = !rancherSemVer.LessThan(minSemVer) &&
```

Reply:
In my opinion the one I have is longer, but it's much easier to read since you can almost read it like a sentence in plain english, whereas your suggestion cleans it up at the expense of readability. If you don't agree I can change it, but I figured you should know why I chose one over the other first.